File: leetcode/array/3.py
from typing import List


class Solution:

    def findRepeatNumber(self, nums: List[int]) -> int:
        '''原地交换'''
        for i, num in enumerate(nums):
            while i != num:  #当前位索引!=当前位
                if num == nums[num]:  #如果已经归位(nums[num])
                    return num
                nums[i], nums[num] = nums[num], nums[i]  #归位(nums[num])
                num = nums[i]
        return -1

    def findRepeatNumber_del(self, nums: List[int]) -> int:
        '''内存占用高'''
        dic = {}
        for i in nums:
            if i not in dic:
                dic[i] = True
            else:
                return i
        return -1


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    solution = Solution()
    print(solution.findRepeatNumber(arr))
    print(arr)

File: leetcode/array/3.cpp
#include <iostream>

class Solution {
 public:
  int find_repeat_number(int arr[], int len) {
    for (int i = 0; i < len; ++i) {
      while (i != arr[i]) {
        if (arr[i] == arr[arr[i]]) {
          return arr[i];
        } else {
          std::swap(arr[i], arr[arr[i]]);
        }
      }
    }
    return -1;
  }
};

int main() {
  int a[6] = {1, 5, 0, 2, 2, 3};

  Solution solution = Solution();
  std::cout << "repeat num: " << solution.find_repeat_number(a, 6) << std::endl;
}